docs/source/development/controllers.rst

```diff
@@ -0,0 +1,221 @@
+can read the state and history of the eVolver environment (e.g. the sensor
+values) and subsequently modify it (via effectors). In a typical eVolver setup
+controller logic is executed once during a device loop, after reading the state
+of all sensors. See :doc:`/concepts` for more details on the device
+operation.
+
+Controllers are implemented as Python classes that inherit from the
+:class:`evolver.controller.Controller` class, the logic of which is executed via
+a call to the :meth:`evolver.controller.Controller.control` method. Developing a
+new controller is a matter of implementing this method. Within
+:meth:`evolver.controller.Controller.control`, code can read sensor values (e.g.
+by calling :meth:`evolver.hardware.interface.SensorDriver.get`) and effect the
+environment based on read sensor values (e.g. by calling
+:meth:`evolver.hardware.interface.EffectorDriver.set` with appropriate input).
+
+As with other eVolver components, controllers contain `Config` classes that
+define the configuration options for the experiment, which can be read from
+on-disk config and modified via the API and UIs. authors should provide a
+`Config` class that inherits from :class:`evolver.controller.Controller.Config`
+with appropriate options for the experiment. See :doc:`config` for
+more details on configuration.
+
+Getting started
+---------------
+
+A breif example will help illustrate what is required to implement a new
+controller. In the example below, we define a simple experiment::
+
+    from evolver.controller import Controller
+    from evolver.hardware.standard import Temperature
+
+    class MyController(Controller):
+        class Config(Controller.Config):
+            limit: int = 42
+            hysteresis: int = 2
+
+        def control(self):
+            for vial in self.vials:
+                density = self.evolver.hardware['temp'].get[vial].density
+                if density > self.limit:
+                    self.evolver.hardware['temp'].set(Temperature.Input(vial=vial, value=30))
+                elif density < self.limit - self.hysteresis:
+                    self.evolver.hardware['temp'].set(Temperature.Input(vial=vial, value=38))
+
+In the above you can note that all the business logic is contained within the
+`control` method, while configuration parameters are available as attributes of
+the controller (e.g. `self.limit` and `self.hysteresis`). Further, the
+controller gains access to hardware state via the `evolver` property, which is
+an instance of :py:class:`evolver.device.Evolver`, and can be used to access
+hardware drivers (e.g. `self.evolver.hardware['temp']`) to get current state and
+set desired state.
+
+.. note::
+
+    The `evolver` property comes from hookup to the device (which happens for
+    example in creating a new eVolver from configuration), see also
+    :ref:`evolver-hookup` below. `vials` comes from the base controller `Config`
+    class, and represents a list of vials to operate on.
+
```

Comment:
I think we need to emphasize this aspect constantly throughout. It may require some significant user/developer understanding and deviation from a more intuitive procedural control method.